marimo/_islands/island_generator.py

```diff
@@ -0,0 +1,189 @@
+# Copyright 2024 Marimo. All rights reserved.
+from __future__ import annotations
+
+from textwrap import dedent
+from typing import cast
+
+from marimo import __version__, _loggers
+from marimo._ast.app import App, InternalApp
+from marimo._ast.cell import Cell, CellConfig
+from marimo._ast.compiler import compile_cell
+from marimo._output.utils import uri_encode_component
+from marimo._server.export.utils import run_app_until_completion
+from marimo._server.file_manager import AppFileManager
+
+LOGGER = _loggers.marimo_logger()
+
+
+class MarimoIslandGenerator:
+    """
+    Generates Marimo islands for embedding in other pages.
+
+    This is a great way to use another SSG framework that converts
+    Python code to HTML using marimo-islands.
+
+    Generally you will want to:
+
+    1. Find all the code snippets and add them to the generator.
+    2. Build the app.
+    3. Replace all code snippets with the rendered HTML.
+    4. Include the header in the <head> tag.
+
+    # Example
+
+    ```python
+    from marimo import MarimoIslandGenerator
+
+    generator = MarimoIslandGenerator()
+    generator.add_code("import marimo as mo")
+    generator.add_code("mo.md('Hello, islands!')")
+
+    # Build the app
+    app = await generator.build()
+
+    # Render the app
+    output = f\"\"\"
+    <html>
+        <head>
+            {generator.render_header()}
+        </head>
+        <body>
+            {generator.render("import marimo as mo")}
+            {generator.render("mo.md('Hello, islands!')")}
+        </body>
+    </html>
+    \"\"\"
+    ```
+    """
```

Comment:
my tests/examples where not the best since i did define the code for `add_cell`, but in reality, i expected it to mostly be opaque and come from the parser as just an unknown snippet, and then output using the same snippet. and wanted to avoid ids and force you to keep around state.

also did not do `render_body()`, or `.body` since i figured the cells would be mixed in with the doc (but i could add this later as well). 

i like returning the stub though with a render function on it (and optional params), i think that is a nicer API. and i can add the `_repr_` as well